File: src/leetcode/python/easy/LowestCommonAncestorOfABinarySearchTree_235.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class LowestCommonAncestorOfABinarySearchTree_235(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if root is None:
            return None
        if p and not q:
            return p
        if not p and q:
            return q
        if p.val < root.val > q.val:
            return self.lowestCommonAncestor(root.left, p, q)
        elif p.val > root.val < q.val:
            return self.lowestCommonAncestor(root.right, p, q)
        return root

File: src/leetcode/java/LowestCommonAncestorOfABinarySearchTree_235.java
package leetcode.java;

public class LowestCommonAncestorOfABinarySearchTree_235 {
  public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    if (root == null) return null;
    if (p == null && q != null) return q;
    if (p != null && q == null) return p;
    if (p.val < root.val && root.val < q.val || q.val < root.val && root.val < p.val) return root;
    if (p.val == root.val || q.val == root.val) return root;
    if (p.val < root.val && q.val < root.val) return lowestCommonAncestor(root.left, p, q);
    return lowestCommonAncestor(root.right, p, q);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
